r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -441,6 +402,43 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(r_gadget)
     }
 
+    /// Reducing a NonNativeMulResultGadget back to a non-native field gadget
+    /// in normal form.
+    pub fn reduce<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<NonNativeFieldGadget<SimulationF, ConstraintF>, SynthesisError> {
+        if super::is_pseudo_mersenne::<SimulationF>() {
+            let params = get_params_for_pseudomersenne(
+                SimulationF::size_in_bits(),
+                ConstraintF::size_in_bits(),
+                SimulationF::Params::DIFFERENCE_WITH_HIGHER_POWER_OF_TWO.unwrap(),
+            );
+            // in this case the first num_limbs-1 limbs should be the constant zero
+            for i in 0..params.num_limbs - 1 {
+                assert_eq!(self.limbs[i].get_value().unwrap(), ConstraintF::zero())
+            }
+            // So we can safely convert the `num_limbs` least significant limbs to a NonNativeFieldGadget
+            // and then employ the more efficient reduce operation
+            let limbs = self.limbs[params.num_limbs - 1..].iter().cloned().collect();
+            // num_add of the NonNativeFieldGadget is 2^bits_per_limb times higher than the num_add of
+            // NonNativeFieldMulResultGadget, as the limbs in normal form are bits_per_limb smaller
+            let num_of_additions_over_normal_form = (&self.num_add_over_normal_form
+                + BigUint::one())
+                * BigUint::from(2usize).pow(params.bits_per_limb as u32)
+                - BigUint::one();
+            let mut result = NonNativeFieldGadget {
+                limbs,
+                num_of_additions_over_normal_form,
+                simulation_phantom: PhantomData,
+            };
+            Reducer::reduce(&mut cs, &mut result)?;
+            return Ok(result);
```

Comment:
Why don't we put this in an extra function, as we did in other places too?

Reply:
I avoid putting it in an ad-hoc function as it is just a simple code that coverts a `NonNativeFieldMulResultGadget` to `NonNativeFieldGadget` and then call the reduce, I thought it was improper to add a function named like `reduce_for_pseudo-mersenne`, as most of the reduction is done with the standard `reduce` function. Do you prefer to have this code in an extra function?

Reply:
Let us keep it as it is for now. I believe that we will refactor our treatment of discriminating generic/special prime cases anyway, when we drop the mulresult gadget. As mentioned below, let us add it to the list of our refactoring issue #152.